lib/instream/connection.ex

```diff
@@ -17,35 +17,49 @@ defmodule Instream.Connection do
         auth:      [ method: :basic, username: "root", password: "root" ]
         hosts:     [ "primary.example.com", "secondary.example.com" ],
         http_opts: [ insecure: true, proxy: "http://company.proxy" ],
+        loggers:   [{ LogModule, :log_fun, [ :additional, :args ] }],
         password:  "pass",
         pool:      [ max_overflow: 10, size: 5 ],
         port:      8086,
         scheme:    "http",
         username:  "user"
   """
 
+  alias Instream.Log
   alias Instream.Query
   alias Instream.Query.Builder
 
 
+  @type log_entry :: Log.PingEntry.t |Log.QueryEntry.t |
+                     Log.StatusEntry.t | Log.WriteEntry.t
   @type query_type :: Builder.t | Query.t | String.t
 
 
   defmacro __using__(otp_app: otp_app) do
-    quote do
-      @before_compile unquote(__MODULE__)
+    quote bind_quoted: [ otp_app: otp_app ] do
+      @before_compile Instream.Connection
 
       alias Instream.Connection
       alias Instream.Connection.QueryPlanner
       alias Instream.Data
       alias Instream.Pool
       alias Instream.Query
 
-      @behaviour unquote(__MODULE__)
-      @otp_app   unquote(otp_app)
+      @behaviour Connection
+      @otp_app   otp_app
       @config    Connection.Config.config(@otp_app, __MODULE__)
 
-      def __pool__, do: __MODULE__.Pool
+      loggers = Enum.reduce(@config[:loggers], quote(do: entry),
```

Comment:
That would probably work also, but I deliberately took the same route as Ecto here. They also have the callback spec'ed to return the log entry after handling.

If one wants to do nasty things you could probably misuse a custom logger to manipulate the entry for all later loggers... not advisable but at least possible.